File: js/src/ContentTypes.ts
import { ContentType } from './ContentType'
import { Marshaler } from './Marshaler'
import { jsonMarshaler, nullMarshaler } from './marshalers'

const make = <Output, Input>(name: string, marshaler: Marshaler<Output, Input>): ContentType<Output, Input> =>
  ({name, marshaler})

export const ContentTypes = {
  json:   make('application/json', jsonMarshaler),
  binary: make('application/octet-stream', nullMarshaler)
}

File: swift/Sources/InterOp/ContentTypes.swift
import Foundation

struct ContentTypes {
    static let standard = ContentTypes(
        ContentType.json,
        ContentType.binary
    )
    
    private var types: [ContentType]
    
    init(_ types: ContentType...) {
        self.types = types
    }
    
    mutating func register(_ contentTypeName: String, _ marshaler: Marshaler) -> ContentType {
        let contentType = ContentType(contentTypeName, marshaler)
        types.append(contentType)
        return contentType
    }
    
    func findByName(_ contentTypeName: String) -> ContentType? {
        return types.first { $0.name == contentTypeName }
    }
}
